File: go/d_store.go
package futures

import "sync/atomic"

type DStoreId struct {
    id *int64
}

func NewDStoreId(id int64) DStoreId {
    return DStoreId{id: &id}
}

func (d DStoreId) Clone() DStoreId {
    newId := atomic.LoadInt64(d.id)
    return NewDStoreId(newId)
}

File: rust/src/d_store.rs
use std::sync::Arc;

#[derive(Debug, Clone)]
pub struct DStoreId(Arc<u64>);

impl DStoreId {
    pub fn new(id: u64) -> Self {
        DStoreId(Arc::new(id))
    }

    pub fn clone(&self) -> Self {
        DStoreId(Arc::clone(&self.0))
    }
}